storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3577,24 +3580,339 @@ dberr_t fsp_sys_tablespace_validate()
 }
 #endif /* UNIV_DEBUG */
 
-void fsp_system_tablespace_truncate()
+/** Store the inode information which basically stores
+the page and offset */
+struct inode_info : private std::unordered_set<uint64_t>
+{
+public:
+  __attribute__((warn_unused_result))
+  /** Register an inode */
+  bool insert_seg(const byte *inode)
+  {
+    uint32_t page= mach_read_from_4(inode + 4);
+    uint32_t offset= mach_read_from_2(inode + 8);
+    return page >= fil_system.sys_space->free_limit ||
+      offset < FIL_PAGE_DATA || offset >= srv_page_size - FIL_PAGE_DATA_END ||
+      emplace(uint64_t{page} << 16 | offset).second;
+  }
+
+  __attribute__((warn_unused_result))
+  bool find(uint32_t page, uint16_t offset) const
+  { return std::unordered_set<uint64_t>::find(uint64_t{page} << 16 | offset) != end(); }
+};
+
+/** Get the file segments from root page
+@param inodes store the index nodes information
+@param root   root page
+@return error code */
+static dberr_t fsp_table_inodes_root(inode_info *inodes, uint32_t root)
+{
+  if (root == FIL_NULL)
+    return DB_SUCCESS;
+
+  dberr_t err= DB_SUCCESS;
+  buf_block_t *block= buf_pool.page_fix(page_id_t{0, root}, &err,
+                                        buf_pool_t::FIX_WAIT_READ);
+  if (!block)
+    return err;
+
+  if (!inodes->insert_seg(block->page.frame + PAGE_HEADER + PAGE_BTR_SEG_TOP))
+    return DB_CORRUPTION;
+
+  if (!inodes->insert_seg(block->page.frame + PAGE_HEADER + PAGE_BTR_SEG_LEAF))
+    return DB_CORRUPTION;
+  block->page.unfix();
+  return err;
```

Comment:
The two `return` must be replaced with `err= DB_CORRUPTION` assignments. We must not leave the page fixed forever.